File: decorator -pattern/js/src/ShapeDecorator.js
import { Shape } from './Shape.js'

// 抽象装饰类，是否继承Shape可选
export class ShapeDecorator extends Shape {
  constructor(decoratedShape) {
    super()
    this.decoratedShape = decoratedShape
  }

  draw() {
    this.decoratedShape.draw()
  }
}

File: decorator -pattern/java/src/ShapeDecorator.java
package src;

// 抽象装饰类，是否实现Shape可选
public abstract class ShapeDecorator implements Shape {
   protected Shape decoratedShape;

   public ShapeDecorator(Shape decoratedShape) {
      this.decoratedShape = decoratedShape;
   }

   public void draw() {
      decoratedShape.draw();
   }
}